File: 2024/10/sol.py
import collections
import timeit

with open('input.txt') as f:
    lines = f.read().splitlines()
    grid = [[int(x) if x.isdigit() else x for x in list(l)] for l in lines]
    print(grid)
    Y = len(grid)
    X = len(grid[0])
    heads = []
    for y in range(Y):
        for x in range(X):
            if grid[y][x] == 0:
                heads.append((y, x))
    print(heads)

dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def traverse_head(head):
    q = collections.deque()
    q.append(head)
    tails = set()
    while q:
        curry, currx = q.popleft()
        curr_val = grid[curry][currx]
        print("iter, ", q)
        if curr_val == 9:
            tails.add((curry, currx))
            continue

        for dy, dx in dirs:
            nexty, nextx = curry + dy, currx + dx
            if nexty in range(Y) and nextx in range(X):
                next_val = grid[nexty][nextx]
                if next_val == ".":
                    continue
                if grid[nexty][nextx] == curr_val + 1:
                    q.append((nexty, nextx))


    return len(tails)

def f():
    count = 0
    for head in heads:
        count += traverse_head(head)

    print(count)
    return count


if __name__ == '__main__':
    iters = 1
    x = timeit.timeit(lambda: f(), number=iters)
    print(f"{x / iters: 0.2f} s")

File: 2024/10/sol.ts
type Coord = {
    y: number,
    x: number,
}

const parse_input = (text: string): number[][] => {
    return text.split("\r\n")
        .map(row => row.split("")
            .map(i => Number(i)))
}

const find_heads = (grid: Number[][]): Coord[] => {
    const heads: Coord[] = []
    for (const [y, row] of grid.entries()) {
        for (const [x, val] of row.entries()) {
            if (val == 0) {
                heads.push({y, x})
            }
        }
    }
    return heads;
}

const traverse_head = (head: Coord, grid: number[][]) => {
    const dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]];

    const seen_paths: Set<Coord[]> = new Set();
    const q: Coord[][] = [[head]]
    while (q.length > 0) {
        const curr_path = q.shift()! // I know this is N
        const {y, x} = curr_path.at(-1)!
        const curr_val = grid[y][x];
        if (curr_val == 9) {
            seen_paths.add(curr_path)
            continue;
        }

        for (const [dy, dx] of dirs) {
            const next_y = y + dy, next_x = x + dx;
            if (0 <= next_y && next_y < grid.length &&
                0 <= next_x && next_x < grid[0].length) {
                const next_val = grid[next_y][next_x];
                if (next_val == curr_val + 1) {
                    const next_path = curr_path.slice();
                    next_path.push({y: next_y, x: next_x});
                    q.push(next_path)
                }
            }
        }
    }
    return seen_paths.size;
}


const grid = parse_input(await Bun.file("./2024/10/input.txt").text());
const heads = find_heads(grid);
const score = heads.reduce((acc, head) => acc + traverse_head(head, grid), 0)
console.log(score);


